File: web/creator/outputs.js
// Where a render lands under ComfyUI/output — the mirror of `outputs.py`.
//
// The duplication is the same deal `canvas.js` has with `canvas.py`: the field
// tells you what is wrong while you are typing, and shows the folder the files
// will actually land in, which cannot wait for a queue round-trip. `outputs.py`
// is authoritative and refuses the same strings at execute time; a hand-edited
// blob never reaches this file at all. `tests/test_outputs_mirror.py` asserts
// the two still agree.

import { t } from "./i18n.js";

// The two shelves everything lands under. Where a *family* lands inside them is
// composed on the Python side and served in its manifest (`output`), because
// the save node and the settings page must show one answer, not two spellings
// of it. These are here for the mirror test, which is what holds the two files
// together at all.
export const RENDERS = "continuity/renders";
export const STILLS = "continuity/stills";

// What core's `compute_vars` expands, in the order the hint lists them. The
// split is real and the chip row draws it: the first six come off the clock at
// the moment the file is written, the last two off the frame being written.
export const CLOCK_TOKENS = ["%year%", "%month%", "%day%", "%hour%", "%minute%", "%second%"];
export const FRAME_TOKENS = ["%width%", "%height%"];
export const TOKENS = [...CLOCK_TOKENS, ...FRAME_TOKENS];

// One token, whole. Used to find them inside typed or pasted text — the field
// draws each one as a single tile, so it has to know where one starts and ends
// rather than treating `%` as a character like any other.
const TOKEN_RE = new RegExp(TOKENS.join("|"), "g");

/** The word a token wears on screen. The `%` is core's syntax, not something a
 *  reader of a folder name should have to look at. */
export function tokenLabel(token) {
  return token.replaceAll("%", "");
}

/**
 * What every token expands to right now.
 *
 * One table, because the preview line under the field and the chip that inserts
 * the token have to agree about whether `%month%` is `08` or `August`, and a
 * chip that says one thing while the line below says another is worse than no
 * chip at all.
 */
export function tokenValues({ width = 1344, height = 768, now = new Date() } = {}) {
  const pad = (n) => String(n).padStart(2, "0");
  return {
    "%year%": String(now.getFullYear()),
    "%month%": pad(now.getMonth() + 1),
    "%day%": pad(now.getDate()),
    "%hour%": pad(now.getHours()),
    "%minute%": pad(now.getMinutes()),
    "%second%": pad(now.getSeconds()),
    "%width%": String(width),
    "%height%": String(height),
  };
}

/**
 * A prefix split into the literal text and the whole tokens inside it, in order.
 *
 * `[{ text }]` and `[{ token }]` parts, never a part that is both. This is what
 * lets the field draw a token as one object: the caret can sit either side of a
 * tile but there is no position *inside* one, which is the whole reason
 * `minima%sssyear%%month%x` was typeable at all.
 */
export function splitTokens(text) {
  const parts = [];
  let at = 0;
  for (const match of String(text ?? "").matchAll(TOKEN_RE)) {
    if (match.index > at) parts.push({ text: text.slice(at, match.index) });
    parts.push({ token: match[0] });
    at = match.index + match[0].length;
  }
  if (at < text.length) parts.push({ text: text.slice(at) });
  return parts;
}

// Windows cannot create these, and a workflow shared from a Mac should not
// produce a folder its Windows half cannot write into. Mirrors outputs.ILLEGAL.
// eslint-disable-next-line no-control-regex
const ILLEGAL = /[<>:"|?*\u0000-\u001f]/;

/**
 * A typed prefix -> `{prefix}` or `{error}`. Never both.
 *
 * Returning the error rather than throwing it is what lets the field show the
 * reason under a value the user is still editing instead of clearing it.
 *
 * @param {string} raw
 * @param {string} fallback   the node's default, used when the field is empty
 */
export function cleanPrefix(raw, fallback) {
  const text = String(raw ?? "").trim().replace(/\\/g, "/");
  if (!text) return { prefix: fallback };

  if (text.startsWith("/") || /^[A-Za-z]:/.test(text)) {
    return { error: t("Absolute paths are not allowed — this is relative to ComfyUI's "
                    + "output folder. Use --output-directory to move that.") };
  }

  // A trailing slash names a folder; keep the default's filename stem.
  const whole = text.endsWith("/") ? text + fallback.split("/").pop() : text;

  const parts = whole.split("/");
  for (const part of parts) {
    if (!part) return { error: t("Empty folder name.") };
    if (part === "." || part === "..") return { error: t("'.' and '..' are not allowed.") };
    if (part.startsWith(".")) return { error: t("Folder names cannot start with a dot.") };
    if (ILLEGAL.test(part)) return { error: t('A name cannot contain any of < > : " | ? *') };
    if (part.endsWith(" ") || part.endsWith(".")) {
      return { error: t("A name cannot end with a space or a dot.") };
    }
  }
  return { prefix: parts.join("/") };
}

/** The folder half of a prefix — what the gallery will file the result under.
 *  "" means the root of the output folder. */
export function folderOf(prefix) {
  const cut = prefix.lastIndexOf("/");
  return cut === -1 ? "" : prefix.slice(0, cut);
}

/** The filename half: the stem every file in that folder is numbered off. */
export function stemOf(prefix) {
  return prefix.slice(prefix.lastIndexOf("/") + 1);
}

/**
 * What the first file will be called, with the tokens filled in from the clock
 * — the whole point of the preview line under the field, since `%month%` on its
 * own does not tell anyone whether they are getting `08` or `August`.
 *
 * The counter is core's, per folder and per stem, so `_00001_` is only right
 * for the first render into an empty one. Shown as an example, labelled as one.
 */
export function examplePath(prefix, { extension = "mp4", width = 1344, height = 768 } = {}) {
  const values = tokenValues({ width, height });
  const filled = prefix.replaceAll(TOKEN_RE, (token) => values[token]);
  return `${filled}_00001_.${extension}`;
}

File: creator/outputs.py
"""Where a render lands under ComfyUI/output.

Both save nodes take a `filename_prefix` and hand it to core's
`get_save_image_path`, which splits it into a subfolder and a filename stem,
expands `%year%`-style tokens, and refuses anything that would write outside the
output directory. That was always true; what was missing is that the prefix was
a module constant, so nobody could set it.

The prefix is the whole output-structure control. `%date%`-style tokens and
`/`-separated folders come from core for free, so this module only has to say
what the defaults are and turn a typed string into one core will accept —
*before* anything is queued, because `get_save_image_path` raising after a
render has been sampled costs the user the render.

The ComfyUI output directory itself is not ours to move: `--output-directory`
(or `--base-directory`) relocates it for every node pack at once, and every path
here is relative to whatever that resolved to.
"""

import re

# The two shelves. Both live under `continuity/` so the pack keeps one shelf in
# somebody else's output folder, and they are split one level deeper so the
# gallery sorts finished videos apart from PreStage stills with no special case
# in the picker: a shelf chip is just a subfolder, and these are two subfolders.
#
# **These are the default and only the default.** They were `minimax/` until the
# pack stopped being named after one of its four families, and an install that
# already has a settings file keeps whatever is in it — see `docs/RENAME.md`.
# Where somebody's finished renders land is not something a rename gets to
# change under them.
RENDERS = "continuity/renders"
STILLS = "continuity/stills"

# What the two shelves held before there was a second family: one folder each,
# with H3's name on the files in both. Kept because `settings.py` has to be able
# to recognise them — a machine that never typed a folder is on these, and
# migrating it onto the per-family defaults below must not look like a folder
# somebody chose. Nothing else reads them; they are history, not a default.
#
# Spelled out rather than built off `RENDERS`, which is what they used to be:
# these name a folder that exists on somebody's disk, so they have to keep
# saying `minimax/` after the default above stopped.
LEGACY_VIDEO_PREFIX = "minimax/renders/H3"
LEGACY_IMAGE_PREFIX = "minimax/stills/prestage"

# Tokens core's `compute_vars` expands. Listed only so the error message can
# tell the user what *is* allowed when they typo one; core owns the real list.
TOKENS = ("%year%", "%month%", "%day%", "%hour%", "%minute%", "%second%",
          "%width%", "%height%")

# Windows cannot create these, and a workflow shared from a Mac should not
# produce a folder its Windows half cannot write into.
ILLEGAL = re.compile(r'[<>:"|?*\x00-\x1f]')


class PrefixError(ValueError):
    """A typed output prefix cannot be used as a path."""


def clean(raw, default):
    """A user-typed output prefix -> one `get_save_image_path` will accept.

    Raises `PrefixError` rather than repairing: a prefix that has to be
    rewritten to be safe is one the user should see refused, so the files land
    where the field says they will. The exception is a trailing slash, which is
    not a mistake but a folder — "my-project/" plainly means "in there, named
    the usual thing", so it keeps the default's stem.
    """
    text = ("" if raw is None else str(raw)).strip().replace("\\", "/")
    if not text:
        return default

    if text.startswith("/") or re.match(r"^[A-Za-z]:", text):
        raise PrefixError(
            f"output folder {raw!r}: absolute paths are not allowed — the prefix is "
            "relative to ComfyUI's output folder. Use --output-directory to move "
            "the output folder itself.")

    # A trailing slash names a folder; keep the default's filename stem.
    if text.endswith("/"):
        text += default.rsplit("/", 1)[-1]

    parts = text.split("/")
    for part in parts:
        if not part:
            raise PrefixError(f"output folder {raw!r}: empty folder name.")
        if part in (".", ".."):
            raise PrefixError(f"output folder {raw!r}: '.' and '..' are not allowed.")
        if part.startswith("."):
            raise PrefixError(f"output folder {raw!r}: folder names cannot start with a dot.")
        if ILLEGAL.search(part):
            raise PrefixError(
                f"output folder {raw!r}: a name cannot contain any of < > : \" | ? *")
        if part.endswith(" ") or part.endswith("."):
            raise PrefixError(
                f"output folder {raw!r}: a name cannot end with a space or a dot.")
    return "/".join(parts)


def default_video(family, stem):
    """Where `family` files its renders when nobody has said otherwise.

    A folder per family under the renders shelf, named by the family id, and the
    family's own stem on the files in it. Two statements at once and both are
    wanted: the folder is what the gallery's shelf chips sort on, so a piece
    shot on LTX 2.5 is one click away from the H3 ones rather than interleaved
    with them, and the stem is what a file dragged out of that folder still says
    about itself.

    Composed here rather than written out in each `declare.py` so that the shape
    of the tree is one decision — a family declares its name, not its path.
    """
    return f"{RENDERS}/{family}/{stem}"


def default_image(family, stem):
    """Where `family` files its stills. The stills shelf, laid out the same way
    — see `default_video`, of which this is the other half."""
    return f"{STILLS}/{family}/{stem}"


# The blob's own key is no longer written by the UI — where files land is a
# preference of this machine and lives in `settings.py`, so a workflow shared
# with someone else does not carry your folder names to their disk. It is still
# *read*, because a hand-edited blob is a supported way to drive these nodes and
# a prefix typed into one has to mean something. Blob first, then the setting.
#
# `default` is the caller's and is required: it is that family's row of the
# settings file, and there is no pack-wide answer left to fall back to.
def video(data, default):
    """The prefix a Creator or Timeline render lands under."""
    return clean(data.get("output_prefix"), default)


def image(data, default):
    """The prefix a PreStage still lands under."""
    return clean(data.get("output_prefix"), default)


def takes(prefix):
    """A render's prefix -> the prefix its individual passes land under.

    One folder deeper, and named for what is in it. A piece shot a pass at a
    time writes a file per pass as well as the piece, and those are working
    files: they are what a card plays instead of being sampled again, and they
    are of no interest once the piece is finished. In their own shelf they sort
    apart from the renders in the gallery and can be swept in one gesture; mixed
    in with them, a ten-card piece would bury the video it made.

    The stem is the render's own, so a take is recognisably from the same
    render as the file beside it.
    """
    folder, _, stem = prefix.rpartition("/")
    return f"{folder}/takes/{stem}" if folder else f"takes/{stem}"
